cni-repair-controller/src/lib.rs

```diff
@@ -0,0 +1,345 @@
+use futures::{Stream, StreamExt};
+use k8s_openapi::api::core::v1::{ObjectReference, Pod};
+use kube::{
+    api::DeleteParams,
+    runtime::{
+        events::{Event, EventType, Recorder, Reporter},
+        watcher,
+    },
+    Client, Error, Resource, ResourceExt,
+};
+use kubert::Runtime;
+use prometheus_client::{
+    metrics::{counter::Counter, histogram::Histogram},
+    registry::Registry,
+};
+use tokio::sync::mpsc::{self, error::TrySendError, Receiver, Sender};
+use tokio::task::JoinHandle;
+use tokio::time::{self, Duration, Instant};
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+// If the event channel capacity is reached, the event is dropped, but a new one will be emitted
+// in the pod's next crashloop iteration
+const EVENT_CHANNEL_CAPACITY: usize = 32;
+
+const DATA_PLANE_LABEL: &str = "linkerd.io/control-plane-ns";
+const EVENT_ACTION: &str = "Deleting";
+const EVENT_REASON: &str = "LinkerdCNINotConfigured";
+
+#[derive(Clone, Debug)]
+pub struct Metrics {
+    queue_overflow: Counter<u64>,
+    pods_deleted: Counter<u64>,
+    pods_delete_latency_seconds: Histogram,
+    pods_delete_errors: Counter<u64>,
+    pods_delete_timeouts: Counter<u64>,
+    events_publish_latency_seconds: Histogram,
+    events_publish_errors: Counter<u64>,
+    events_publish_timeouts: Counter<u64>,
+}
+
+pub fn run(
+    rt: &mut Runtime,
+    node_name: String,
+    controller_pod_name: String,
+    metrics: Metrics,
+) -> JoinHandle<()> {
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics))
```

Comment:
It's probably desirable to put tracing contexts on each span for better logging/diagnostics:

```suggestion
    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
    tokio::spawn(
        process_events(pod_evts, tx, metrics.clone())
            .instrument(tracing::info_span!("watch").or_current()),
    );

    let client = rt.client();
    tokio::spawn(
        process_pods(client, controller_pod_name, rx, metrics)
            .instrument(tracing::info_span!("repair").or_current()),
    )
```

Reply:
Sounds good :+1: